Source/Internal/HAConnectionImpl.swift

```diff
@@ -401,6 +414,8 @@ extension HAConnectionImpl {
             sendWebSocket(identifier: identifier, request: request, command: command)
         case let .rest(method, command):
             sendRest(identifier: identifier!, request: request, method: method, command: command)
+        case let .sttData(data):
+            sendWrite(data.sttBinaryHandlerId, audioDataString: request.data["audioData"] as? String)
```

Comment:
I'm guessing this isn't invoking the completion blocks, but it _may_ be useful to either wait to invoke it until the write occurs, or just invoke the completion right away. I know the server isn't ACK-ing this but it's still good form to be able to chain these together for throttling if you want.

Reply:
Communicating to the response controller isn't enough? like done here:

```
 private func sendWrite(_ sttBinaryHandlerId: UInt8, audioDataString: String?) {
        // If there is no audioData, handlerID will be the payload alone indicating end of audio
        var audioData = Data(base64Encoded: audioDataString ?? "") ?? Data()

        // Prefix audioData with handler ID so the API can map the binary data
        audioData.insert(sttBinaryHandlerId, at: 0)
        workQueue.async { [connection] in
            connection?.write(data: audioData) { [weak self] in
                self?.responseController.didWrite()
            }
        }
    }
```

I noticed that when "sending" using websocket or rest, that's the only response, completion is not used for those either

Reply:
Not all of the tests check the invocation of the completion handler, but some do. e.g. testPlainSendSentSuccessful, testPlainSendSentFailurePromise, etc. You're making a new one with a unique code path, so you need to make sure all of the same guarantees happen somewhere.

Reply:
Ah, got it, ok I added the identifier and a test to validate it, I also added this code block inside "sendWrite":

```
 if let identifier, let request = self.requestController.single(for: identifier) {
                    callbackQueue.async {
                        request.resolve(.success(.empty))
                    }
                    requestController.clear(invocation: request)
                }
```

Because I can't add identifier to the data dictionary (since I am sending binary in this write) and it will also have no websocket event back from the connection.

Does that make sense?